sigmoid: code cleanup

flannelhead opened the pull request:
Cleanup around the hue preservation code in sigmoid.

* Remove the separate path for case when the hue preservation parameter is near zero
* Reuse some calculation results in the hue preservation code
* Refine handling of special cases / avoiding of NaNs in the hue preservation code. Now there are explicit checks in divisions where the denominator might be zero
* Replace some double-precision number literals with single-precision ones

The main aim was to remove the separate processing path / OpenCL kernel for the hue_preservation = 0 case. I've been discussing the possibility of adding custom tone mapping primaries to the module with @jandren, and this cleanup paves the way for that.

Integration test is clean.
```
Test 0097-sigmoid
      Image mire1.cr2
      CPU & GPU version differ by 22836 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 1.52457
      Avg dE                   : 0.00372
      Std dE                   : 0.04458
      ----------------------------------
      Pixels below avg + 0 std : 99.18 %
      Pixels below avg + 1 std : 99.18 %
      Pixels below avg + 3 std : 99.21 %
      Pixels below avg + 6 std : 99.35 %
      Pixels below avg + 9 std : 99.46 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 1.31436
      Avg dE                   : 0.00012
      Std dE                   : 0.00811
      ----------------------------------
      Pixels below avg + 0 std : 99.97 %
      Pixels below avg + 1 std : 99.97 %
      Pixels below avg + 3 std : 99.97 %
      Pixels below avg + 6 std : 99.97 %
      Pixels below avg + 9 std : 99.97 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
  OK
```

This would probably deserve a review from @jandren.

jandren reviewed (approved):
Thanks flannelhead!

There isn't any shorter code path of early return now for the non hue corrected case. But that might not affect performance much anyway.

Maybe drop "_interpolated" from the function name now? The suffix is kind of redundant now.

flannelhead commented:
Thanks for the review @jandren. Removed the `_interpolated` suffix as requested.

Did some additional testing to ensure everything is alright. Added a new test with the hue preservation set to 0 (expected from the code state before this PR):
```
Test 0147-sigmoid-no-hue-preservation
      Image mire1.cr2
      CPU & GPU version differ by 23546 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 2.13192
      Avg dE                   : 0.00377
      Std dE                   : 0.04475
      ----------------------------------
      Pixels below avg + 0 std : 99.16 %
      Pixels below avg + 1 std : 99.16 %
      Pixels below avg + 3 std : 99.21 %
      Pixels below avg + 6 std : 99.34 %
      Pixels below avg + 9 std : 99.44 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 1.08418
      Avg dE                   : 0.00001
      Std dE                   : 0.00254
      ----------------------------------
      Pixels below avg + 0 std : 100.00 %
      Pixels below avg + 1 std : 100.00 %
      Pixels below avg + 3 std : 100.00 %
      Pixels below avg + 6 std : 100.00 %
      Pixels below avg + 9 std : 100.00 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
```

Also, as I seem to sometimes get a tiny diff to the expected image on my computer even if nothing has changed, I updated the expected image to a locally rendered one (again, from the code before this PR):
```
Test 0097-sigmoid
      Image mire1.cr2
      CPU & GPU version differ by 23053 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 2.13192
      Avg dE                   : 0.00373
      Std dE                   : 0.04458
      ----------------------------------
      Pixels below avg + 0 std : 99.17 %
      Pixels below avg + 1 std : 99.18 %
      Pixels below avg + 3 std : 99.21 %
      Pixels below avg + 6 std : 99.34 %
      Pixels below avg + 9 std : 99.45 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 0.78833
      Avg dE                   : 0.00002
      Std dE                   : 0.00294
      ----------------------------------
      Pixels below avg + 0 std : 100.00 %
      Pixels below avg + 1 std : 100.00 %
      Pixels below avg + 3 std : 100.00 %
      Pixels below avg + 6 std : 100.00 %
      Pixels below avg + 9 std : 100.00 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
```

So, everything seems ok to me here.

flannelhead commented:
> There isn't any shorter code path of early return now for the non hue corrected case. But that might not affect performance much anyway.

That is my assumption as well, although I have not run any perf benchmarks here. This change was more about reducing the code footprint and tidying things up. Perhaps also helping the compiler a tiny bit as one branching point (early exit from the hue preservation function) was removed.

TurboGit reviewed (approved):
Thanks!
